yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClusterSchedulerBackend.scala

```diff
@@ -47,4 +47,12 @@ private[spark] class YarnClusterSchedulerBackend(
       super.applicationId
     }
 
+  override def applicationAttemptId(): Option[String] =
+    // In YARN Cluster mode, spark.yarn.app.attemptid is expect to be set
+    // before user application is launched.
+    // So, if spark.yarn.app.id is not set, it is something wrong.
```

Comment:
is this supposed to be `spark.yarn.app.attemptId` instead of just the `app.id`? Maybe a simpler way to put this is "The attempt ID is always set for YARN cluster applications"